### PR TITLE
[Fix #445] def form with strings and docstrings

### LINKED ISSUE #445: (def s "a string") font locks "a string" as a docstring
## Expected behavior

Font lock the string `"a string"` as a string in `(def s "a string")`.

## Actual behavior

The string `"a string"` is font locked as a docstring in `(def s "a string")`.

## clojure-mode version information
```
clojure-mode (version 5.7.0-snapshot)
```

## Proposed solution

I played around with this, and it can be solved by adding a check in this `and` https://github.com/clojure-emacs/clojure-mode/blob/master/clojure-mode.el#L898 in the function `clojure-font-lock-syntactic-face-function`. For example
```
(and ...
  (or (save-excursion
        (goto-char startpos)
        (not (= ?\) (char-after (+ startpos (length (sexp-at-point)) 2)))))
      (string= "defprotocol" firstsym)))
```

This, of course, checks for a closing paren after the string, and if so, it is not considered a docstring, except for `defprotocol`, which is the only kind of form that can have a docstring in the last position. Instead of checking for a paren, one could of course check for another element, but it is more involved.

Opinions, suggestions, etc?

### PR DESCRIPTION
**Pull request by carlosgeos**
I took the proposed solution here: https://github.com/clojure-emacs/clojure-mode/issues/445 (thanks @oskarkv) and scoped it to only the def form.

Basically:
```clojure
(def my_string_with_docstring "how to use" "just a string")
```
`how to use` and `just a string` are highlighted differently.

And:
```clojure
(def im_a_string "hello")
```
`hello` is no longer highlighted as a docstring

I did what was said in these comments: https://github.com/clojure-emacs/clojure-mode/issues/445#issuecomment-340689656 and https://github.com/clojure-emacs/clojure-mode/issues/445#issuecomment-341931740. Now it works for defs, defprotocol, ns (whatever the form)...

-----------------

- [x] The commits are consistent with our [contribution guidelines][1].
- [x] You've added tests (if possible) to cover your change(s). Bugfix, indentation, and font-lock tests are extremely important!
- [x] You've run `M-x checkdoc` and fixed any warnings in the code you've written.
- [x] You've updated the changelog (if adding/changing user-visible functionality).
- [ ] You've updated the readme (if adding/changing user-visible functionality).

### CODE REVIEW

**Comment by bbatsov** [comment]
Please add some unit tests and update the change log as well. (This suggestions applies to the other PRs you’ve opened as well)

**Comment by carlosgeos** [comment]
done for this one, the others will come soon

File: test/clojure-mode-indentation-test.el [hunk 1]
```diff
@@ -68,9 +68,11 @@ values of customisable variables."
            (search-forward "|")
            (delete-char -1)
            (clojure-mode)
+           (font-lock-ensure)
            (indent-according-to-mode)
 
-           (should (equal expected-state (buffer-string)))
+           (should (equal expected-state (buffer-substring-no-properties
```
Comment (bbatsov) [inline test/clojure-mode-indentation-test.el]: Why is this change necessary?

Comment (carlosgeos) [reply test/clojure-mode-indentation-test.el]: Because `clojure-in-docstring-p` checks for a font-lock-face, `indent-according-to-mode` works differently (in this case) if font-lock-mode is off, so I wrote `font-lock-ensure` since otherwise there is no font-lock information in the temp-buffer.

If this change wasn't there, this test: 

https://github.com/clojure-emacs/clojure-mode/blob/43e3e65966b8798968016bf0134a726c4a73461e/test/clojure-mode-indentation-test.el#L123-L125

would always pass, both in the current version and my version. To make it fail in the current version (which is expected and wanted), the change had to be made. I found no other way to mimic what is really happening.

But you are right, I'll compare the strings without removing the properties, no reason to do that.

File: test/clojure-mode-indentation-test.el [hunk 2]
```diff
@@ -68,6 +68,7 @@ values of customisable variables."
            (search-forward "|")
            (delete-char -1)
            (clojure-mode)
+           (font-lock-ensure)
```
Comment (bbatsov) [inline test/clojure-mode-indentation-test.el]: How can this affect an indentation test?

Comment (carlosgeos) [reply test/clojure-mode-indentation-test.el]: I think it is because the indentation command behaves differently depending on whether font lock mode is ON or OFF, in the case of docstrings.

This branch is master with the test added: https://github.com/carlosgeos/clojure-mode/tree/temp-branch. The tests are fine, but if `(font-lock-ensure)` is added, the test with the docstring fails (like it should).

Comment (bbatsov) [reply test/clojure-mode-indentation-test.el]: Hmm, that’s odd. Turning on the mode should enable the font locking.

Comment (carlosgeos) [reply test/clojure-mode-indentation-test.el]: Usually yes, but apparently not when using temp-buffer:

https://stackoverflow.com/a/23569337
https://emacs.stackexchange.com/a/44301

**Comment by carlosgeos** [comment]
True, that example is incorrectly font locked. I'll make some modifications.

**Comment by carlosgeos** [comment]
The new solution handles whitespace and newlines better

If everything is ok, I will squash

File: clojure-mode.el [hunk 3]
```diff
@@ -955,7 +955,13 @@ highlighted region)."
                              (setq docelt (1- docelt)))))
                        (and (zerop docelt) (<= (point) startpos)
                             (progn (forward-comment (point-max)) t)
-                            (= (point) (nth 8 state)))))
+                            (= (point) (nth 8 state))))
+                     ;; In a def, at last position is not a docstring
+                     (not (and (string= "def" firstsym)
+                               (save-excursion
+                                 (goto-char startpos)
```
Comment (cichli) [inline clojure-mode.el]: Is this necessary given the `goto-char` below?

**Comment by cichli** [comment]
Apart from that one final comment, everything looks great now.

File: clojure-mode.el (hunk 3 above)
Comment (carlosgeos) [reply clojure-mode.el]: The first `goto-char` looks redundant but I think it is not, because there is a `(sexp-at-point)` below, and it needs the point at the correct position beforehand.

Also, the `(sexp-at-point)` form does not take any arguments so I cannot pass `startpos` to it.

Anyway, I tried without it and it didn't work. Happy to remove it if we find the way to make it work.

Comment (cichli) [reply clojure-mode.el]: Ah, great point. I didn't think of that! Yeah, we definitely need it because of the `(sexp-at-point)`.

Comment (cichli) [reply clojure-mode.el]: Go ahead and squash if you like. I will merge once I have the rights to do so.

**Comment by carlosgeos** [comment]
Thanks, I also think you can close https://github.com/clojure-emacs/clojure-mode/issues/469 and https://github.com/clojure-emacs/clojure-mode/issues/405
